File: src/test/kotlin/school/charset/app/domain/exercise/generator/CodePointGeneratorTest.kt
package school.charset.app.domain.exercise.generator

import io.kotest.core.spec.style.FreeSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import school.charset.app.domain.encoding.CodePoint
import kotlin.random.Random

class CodePointGeneratorTest :
    FreeSpec({
        // Random is mocked with MockK. Each test pins both the bounds passed to
        // `Random.nextInt(from, until)` (via `every` with explicit arguments) and
        // the value returned. If the production code passes wrong bounds, MockK
        // throws "no mocking found" and the test fails fast.

        "randomAscii" - {
            "Printable - picks from [0x20, 0x7F) = printable ASCII" {
                val random = mockk<Random>()
                every { random.nextInt(0x20, 0x7F) } returns 0x41
                CodePointGenerator(random).randomAscii(AsciiLevel.Printable) shouldBe CodePoint(0x41)
            }

            "Printable - low boundary 0x20" {
                val random = mockk<Random>()
                every { random.nextInt(0x20, 0x7F) } returns 0x20
                CodePointGenerator(random).randomAscii(AsciiLevel.Printable) shouldBe CodePoint(0x20)
            }

            "Printable - high boundary 0x7E" {
                val random = mockk<Random>()
                every { random.nextInt(0x20, 0x7F) } returns 0x7E
                CodePointGenerator(random).randomAscii(AsciiLevel.Printable) shouldBe CodePoint(0x7E)
            }

            "Full - picks from [0x00, 0x80) = full ASCII including controls" {
                val random = mockk<Random>()
                every { random.nextInt(0x00, 0x80) } returns 0x41
                CodePointGenerator(random).randomAscii(AsciiLevel.Full) shouldBe CodePoint(0x41)
            }

            "Full - low boundary 0x00 (NUL)" {
                val random = mockk<Random>()
                every { random.nextInt(0x00, 0x80) } returns 0x00
                CodePointGenerator(random).randomAscii(AsciiLevel.Full) shouldBe CodePoint(0x00)
            }

            "Full - high boundary 0x7F (DEL)" {
                val random = mockk<Random>()
                every { random.nextInt(0x00, 0x80) } returns 0x7F
                CodePointGenerator(random).randomAscii(AsciiLevel.Full) shouldBe CodePoint(0x7F)
            }
        }

        "randomLatin1" - {
            "Supplement - picks from [0xA0, 0x100) = Latin-1 supplement only" {
                val random = mockk<Random>()
                every { random.nextInt(0xA0, 0x100) } returns 0xE9
                CodePointGenerator(random).randomLatin1(Latin1Level.Supplement) shouldBe CodePoint(0xE9)
            }

            "Supplement - low boundary 0xA0 (NBSP)" {
                val random = mockk<Random>()
                every { random.nextInt(0xA0, 0x100) } returns 0xA0
                CodePointGenerator(random).randomLatin1(Latin1Level.Supplement) shouldBe CodePoint(0xA0)
            }

            "Supplement - high boundary 0xFF (ÿ)" {
                val random = mockk<Random>()
                every { random.nextInt(0xA0, 0x100) } returns 0xFF
                CodePointGenerator(random).randomLatin1(Latin1Level.Supplement) shouldBe CodePoint(0xFF)
            }

            "Full - picks from [0x00, 0x100) = full Latin-1" {
                val random = mockk<Random>()
                every { random.nextInt(0x00, 0x100) } returns 0xE9
                CodePointGenerator(random).randomLatin1(Latin1Level.Full) shouldBe CodePoint(0xE9)
            }

            "Full - low boundary 0x00 (NUL)" {
                val random = mockk<Random>()
                every { random.nextInt(0x00, 0x100) } returns 0x00
                CodePointGenerator(random).randomLatin1(Latin1Level.Full) shouldBe CodePoint(0x00)
            }

            "Full - high boundary 0xFF (ÿ)" {
                val random = mockk<Random>()
                every { random.nextInt(0x00, 0x100) } returns 0xFF
                CodePointGenerator(random).randomLatin1(Latin1Level.Full) shouldBe CodePoint(0xFF)
            }
        }

        "randomWindows1252" - {
            // SpecialBlock picks from the 27 special code points only.
            // Indices are stable: 0 = byte 0x80 (€), ..., 26 = byte 0x9F (Ÿ).

            "SpecialBlock - index 0 picks Euro (U+20AC, byte 0x80)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 27) } returns 0
                CodePointGenerator(random).randomWindows1252(Windows1252Level.SpecialBlock) shouldBe
                    CodePoint(0x20AC)
            }

            "SpecialBlock - index 11 picks Œ (U+0152, byte 0x8C)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 27) } returns 11
                CodePointGenerator(random).randomWindows1252(Windows1252Level.SpecialBlock) shouldBe
                    CodePoint(0x0152)
            }

            "SpecialBlock - index 26 picks Ÿ (U+0178, byte 0x9F, last special)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 27) } returns 26
                CodePointGenerator(random).randomWindows1252(Windows1252Level.SpecialBlock) shouldBe
                    CodePoint(0x0178)
            }

            // AllEncodable picks from all 251 encodable code points. Layout:
            // - indices 0..127     : ASCII range (U+0000..U+007F)
            // - indices 128..154   : special block (27 entries, in byte order)
            // - indices 155..250   : Latin-1 supplement (U+00A0..U+00FF)

            "AllEncodable - index 0 picks U+0000 (NUL, start of ASCII)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 251) } returns 0
                CodePointGenerator(random).randomWindows1252(Windows1252Level.AllEncodable) shouldBe
                    CodePoint(0x00)
            }

            "AllEncodable - index 127 picks U+007F (DEL, end of ASCII)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 251) } returns 127
                CodePointGenerator(random).randomWindows1252(Windows1252Level.AllEncodable) shouldBe
                    CodePoint(0x7F)
            }

            "AllEncodable - index 128 picks U+20AC (Euro, first special)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 251) } returns 128
                CodePointGenerator(random).randomWindows1252(Windows1252Level.AllEncodable) shouldBe
                    CodePoint(0x20AC)
            }

            "AllEncodable - index 154 picks U+0178 (Ÿ, last special)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 251) } returns 154
                CodePointGenerator(random).randomWindows1252(Windows1252Level.AllEncodable) shouldBe
                    CodePoint(0x0178)
            }

            "AllEncodable - index 155 picks U+00A0 (NBSP, start of Latin-1 supplement)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 251) } returns 155
                CodePointGenerator(random).randomWindows1252(Windows1252Level.AllEncodable) shouldBe
                    CodePoint(0xA0)
            }

            "AllEncodable - index 250 picks U+00FF (ÿ, end)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 251) } returns 250
                CodePointGenerator(random).randomWindows1252(Windows1252Level.AllEncodable) shouldBe
                    CodePoint(0xFF)
            }
        }

        "randomUtf8" - {
            "OneByte - picks from [0x00, 0x80) = 1-byte UTF-8 (ASCII subset)" {
                val random = mockk<Random>()
                every { random.nextInt(0x00, 0x80) } returns 0x41
                CodePointGenerator(random).randomUtf8(Utf8Level.OneByte) shouldBe CodePoint(0x41)
            }

            "OneByte - low boundary 0x00" {
                val random = mockk<Random>()
                every { random.nextInt(0x00, 0x80) } returns 0x00
                CodePointGenerator(random).randomUtf8(Utf8Level.OneByte) shouldBe CodePoint(0x00)
            }

            "OneByte - high boundary 0x7F" {
                val random = mockk<Random>()
                every { random.nextInt(0x00, 0x80) } returns 0x7F
                CodePointGenerator(random).randomUtf8(Utf8Level.OneByte) shouldBe CodePoint(0x7F)
            }

            "TwoByte - picks from [0x80, 0x800) = 2-byte UTF-8" {
                val random = mockk<Random>()
                every { random.nextInt(0x80, 0x800) } returns 0xE9
                CodePointGenerator(random).randomUtf8(Utf8Level.TwoByte) shouldBe CodePoint(0xE9)
            }

            "TwoByte - low boundary 0x80" {
                val random = mockk<Random>()
                every { random.nextInt(0x80, 0x800) } returns 0x80
                CodePointGenerator(random).randomUtf8(Utf8Level.TwoByte) shouldBe CodePoint(0x80)
            }

            "TwoByte - high boundary 0x7FF" {
                val random = mockk<Random>()
                every { random.nextInt(0x80, 0x800) } returns 0x7FF
                CodePointGenerator(random).randomUtf8(Utf8Level.TwoByte) shouldBe CodePoint(0x7FF)
            }

            // ThreeByte: 3-byte range with surrogate gap. The implementation indexes
            // into a virtual range of 61440 non-surrogate code points:
            // - indices 0..53247   -> U+0800..U+D7FF (before surrogates)
            // - indices 53248..61439 -> U+E000..U+FFFF (after surrogates)

            "ThreeByte - index 0 picks U+0800 (start of 3-byte range)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 61440) } returns 0
                CodePointGenerator(random).randomUtf8(Utf8Level.ThreeByte) shouldBe CodePoint(0x0800)
            }

            "ThreeByte - index 53247 picks U+D7FF (last before surrogates)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 61440) } returns 53247
                CodePointGenerator(random).randomUtf8(Utf8Level.ThreeByte) shouldBe CodePoint(0xD7FF)
            }

            "ThreeByte - index 53248 picks U+E000 (first after surrogates, skip)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 61440) } returns 53248
                CodePointGenerator(random).randomUtf8(Utf8Level.ThreeByte) shouldBe CodePoint(0xE000)
            }

            "ThreeByte - index 61439 picks U+FFFF (BMP max, last of 3-byte range)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 61440) } returns 61439
                CodePointGenerator(random).randomUtf8(Utf8Level.ThreeByte) shouldBe CodePoint(0xFFFF)
            }

            "FourByte - picks from [0x10000, 0x110000) = 4-byte UTF-8 (supplementary plane)" {
                val random = mockk<Random>()
                every { random.nextInt(0x10000, 0x110000) } returns 0x1F600
                CodePointGenerator(random).randomUtf8(Utf8Level.FourByte) shouldBe CodePoint(0x1F600)
            }

            "FourByte - low boundary 0x10000" {
                val random = mockk<Random>()
                every { random.nextInt(0x10000, 0x110000) } returns 0x10000
                CodePointGenerator(random).randomUtf8(Utf8Level.FourByte) shouldBe CodePoint(0x10000)
            }

            "FourByte - high boundary 0x10FFFF" {
                val random = mockk<Random>()
                every { random.nextInt(0x10000, 0x110000) } returns 0x10FFFF
                CodePointGenerator(random).randomUtf8(Utf8Level.FourByte) shouldBe CodePoint(0x10FFFF)
            }
        }

        "randomUtf16" - {
            // Bmp: 63488 non-surrogate code points indexed virtually:
            // - indices 0..55295    -> U+0000..U+D7FF (before surrogates)
            // - indices 55296..63487 -> U+E000..U+FFFF (after surrogates)

            "Bmp - index 0 picks U+0000 (NUL, start of BMP)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 63488) } returns 0
                CodePointGenerator(random).randomUtf16(Utf16Level.Bmp) shouldBe CodePoint(0x0000)
            }

            "Bmp - index 55295 picks U+D7FF (last before surrogates)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 63488) } returns 55295
                CodePointGenerator(random).randomUtf16(Utf16Level.Bmp) shouldBe CodePoint(0xD7FF)
            }

            "Bmp - index 55296 picks U+E000 (first after surrogates, skip)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 63488) } returns 55296
                CodePointGenerator(random).randomUtf16(Utf16Level.Bmp) shouldBe CodePoint(0xE000)
            }

            "Bmp - index 63487 picks U+FFFF (BMP max)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 63488) } returns 63487
                CodePointGenerator(random).randomUtf16(Utf16Level.Bmp) shouldBe CodePoint(0xFFFF)
            }

            "Supplementary - picks from [0x10000, 0x110000) = supplementary plane" {
                val random = mockk<Random>()
                every { random.nextInt(0x10000, 0x110000) } returns 0x1F600
                CodePointGenerator(random).randomUtf16(Utf16Level.Supplementary) shouldBe CodePoint(0x1F600)
            }

            "Supplementary - low boundary 0x10000" {
                val random = mockk<Random>()
                every { random.nextInt(0x10000, 0x110000) } returns 0x10000
                CodePointGenerator(random).randomUtf16(Utf16Level.Supplementary) shouldBe CodePoint(0x10000)
            }

            "Supplementary - high boundary 0x10FFFF" {
                val random = mockk<Random>()
                every { random.nextInt(0x10000, 0x110000) } returns 0x10FFFF
                CodePointGenerator(random).randomUtf16(Utf16Level.Supplementary) shouldBe CodePoint(0x10FFFF)
            }
        }

        "randomUtf32" - {
            // Same backing logic as randomUtf16 (Bmp picks from non-surrogate BMP,
            // Supplementary picks from 0x10000..0x10FFFF). Tests below verify the
            // delegation rather than the underlying math (covered by randomUtf16 tests).

            "Bmp - index 0 picks U+0000 (start of BMP)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 63488) } returns 0
                CodePointGenerator(random).randomUtf32(Utf32Level.Bmp) shouldBe CodePoint(0x0000)
            }

            "Bmp - index 55295 picks U+D7FF (last before surrogates)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 63488) } returns 55295
                CodePointGenerator(random).randomUtf32(Utf32Level.Bmp) shouldBe CodePoint(0xD7FF)
            }

            "Bmp - index 55296 picks U+E000 (first after surrogates, skip)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 63488) } returns 55296
                CodePointGenerator(random).randomUtf32(Utf32Level.Bmp) shouldBe CodePoint(0xE000)
            }

            "Bmp - index 63487 picks U+FFFF (BMP max)" {
                val random = mockk<Random>()
                every { random.nextInt(0, 63488) } returns 63487
                CodePointGenerator(random).randomUtf32(Utf32Level.Bmp) shouldBe CodePoint(0xFFFF)
            }

            "Supplementary - picks from [0x10000, 0x110000)" {
                val random = mockk<Random>()
                every { random.nextInt(0x10000, 0x110000) } returns 0x1F600
                CodePointGenerator(random).randomUtf32(Utf32Level.Supplementary) shouldBe CodePoint(0x1F600)
            }

            "Supplementary - low boundary 0x10000" {
                val random = mockk<Random>()
                every { random.nextInt(0x10000, 0x110000) } returns 0x10000
                CodePointGenerator(random).randomUtf32(Utf32Level.Supplementary) shouldBe CodePoint(0x10000)
            }

            "Supplementary - high boundary 0x10FFFF" {
                val random = mockk<Random>()
                every { random.nextInt(0x10000, 0x110000) } returns 0x10FFFF
                CodePointGenerator(random).randomUtf32(Utf32Level.Supplementary) shouldBe CodePoint(0x10FFFF)
            }
        }
    })

File: test/charset/exercise/generator/code_point_generator_test.exs
defmodule Charset.Exercise.Generator.CodePointGeneratorTest do
  @moduledoc """
  The Kotlin original injected a mocked Random to pin exact draws; here the
  RNG is the process one, so the port asserts by sampling: every draw must
  stay inside the tier's allowed set, and mixed tiers must visit every
  sub-range eventually.
  """
  use ExUnit.Case, async: true

  alias Charset.Encoding.Windows1252
  alias Charset.Exercise.Generator.CodePointGenerator

  @draws 2_000

  defp surrogate?(cp), do: cp in 0xD800..0xDFFF

  describe "ascii" do
    test "printable tier stays in U+0020..U+007E" do
      for _draw <- 1..@draws do
        assert CodePointGenerator.random_ascii(:printable) in 0x20..0x7E
      end
    end

    test "full tier stays in U+0000..U+007F and visits both sub-ranges" do
      samples = for _draw <- 1..@draws, do: CodePointGenerator.random_ascii(:full)

      assert Enum.all?(samples, &(&1 in 0x00..0x7F))
      # 40/60 printable/full mix: control characters must show up
      assert Enum.any?(samples, &(&1 < 0x20))
    end
  end

  describe "latin1" do
    test "supplement tier stays in U+00A0..U+00FF" do
      for _draw <- 1..@draws do
        assert CodePointGenerator.random_latin1(:supplement) in 0xA0..0xFF
      end
    end

    test "full tier stays in U+0000..U+00FF and visits the ASCII range" do
      samples = for _draw <- 1..@draws, do: CodePointGenerator.random_latin1(:full)

      assert Enum.all?(samples, &(&1 in 0x00..0xFF))
      assert Enum.any?(samples, &(&1 < 0xA0))
    end
  end

  describe "windows-1252" do
    test "special block tier draws only the 27 special code points" do
      specials = MapSet.new(Windows1252.special_code_points())

      for _draw <- 1..@draws do
        assert CodePointGenerator.random_windows1252(:special_block) in specials
      end
    end

    test "all-encodable tier draws only encodable code points" do
      encodable = MapSet.new(Windows1252.encodable_code_points())
      samples = for _draw <- 1..@draws, do: CodePointGenerator.random_windows1252(:all_encodable)

      assert Enum.all?(samples, &(&1 in encodable))
      # 30/70 mix: identity ranges must show up alongside the special block
      assert Enum.any?(samples, &(&1 <= 0x7F))
    end
  end

  describe "utf8" do
    test "one-byte tier stays in U+0000..U+007F" do
      for _draw <- 1..@draws do
        assert CodePointGenerator.random_utf8(:one_byte) in 0x00..0x7F
      end
    end

    test "two-byte tier mixes 1-byte (20%) and 2-byte (80%) ranges" do
      samples = for _draw <- 1..@draws, do: CodePointGenerator.random_utf8(:two_byte)

      assert Enum.all?(samples, &(&1 in 0x00..0x7FF))
      assert Enum.any?(samples, &(&1 <= 0x7F))
      assert Enum.any?(samples, &(&1 > 0x7F))
    end

    test "three-byte tier never draws a surrogate" do
      samples = for _draw <- 1..@draws, do: CodePointGenerator.random_utf8(:three_byte)

      assert Enum.all?(samples, &(&1 in 0x00..0xFFFF))
      refute Enum.any?(samples, &surrogate?/1)
      # 10/25/65 mix: the 3-byte sub-range itself must show up
      assert Enum.any?(samples, &(&1 >= 0x800))
    end

    test "four-byte tier spans the whole range, never a surrogate, all 4 sub-ranges visited" do
      samples = for _draw <- 1..@draws, do: CodePointGenerator.random_utf8(:four_byte)

      assert Enum.all?(samples, &(&1 in 0x00..0x10FFFF))
      refute Enum.any?(samples, &surrogate?/1)
      assert Enum.any?(samples, &(&1 <= 0x7F))
      assert Enum.any?(samples, &(&1 in 0x80..0x7FF))
      assert Enum.any?(samples, &(&1 in 0x800..0xFFFF))
      assert Enum.any?(samples, &(&1 >= 0x10000))
    end
  end

  describe "utf16 / utf32" do
    test "bmp tier stays in the BMP and never draws a surrogate" do
      for random <- [&CodePointGenerator.random_utf16/1, &CodePointGenerator.random_utf32/1] do
        samples = for _draw <- 1..@draws, do: random.(:bmp)

        assert Enum.all?(samples, &(&1 in 0x0000..0xFFFF))
        refute Enum.any?(samples, &surrogate?/1)
      end
    end

    test "supplementary tier mixes BMP (30%) and supplementary (70%)" do
      samples = for _draw <- 1..@draws, do: CodePointGenerator.random_utf16(:supplementary)

      assert Enum.all?(samples, &(&1 in 0x0000..0x10FFFF))
      refute Enum.any?(samples, &surrogate?/1)
      assert Enum.any?(samples, &(&1 <= 0xFFFF))
      assert Enum.any?(samples, &(&1 >= 0x10000))
    end
  end

  test "boundary code points are reachable (seeded sweep over the bmp skip logic)" do
    # The surrogate-skip arithmetic must be able to produce both U+D7FF (just
    # before the block) and U+E000 (just after). Sample heavily and check the
    # skip never lands inside the block while both shores stay reachable
    # across the whole BMP sweep.
    samples = for _draw <- 1..50_000, do: CodePointGenerator.random_utf16(:bmp)

    refute Enum.any?(samples, &surrogate?/1)
    assert Enum.any?(samples, &(&1 >= 0xE000))
    assert Enum.any?(samples, &(&1 < 0xD800))
  end
end
